### PR TITLE
Fixed CH4 lock synchronization.

### PR DESCRIPTION
**Pull request by monicasee**
The previous code managed lock epochs per window, thus did not support
simultaneous lock epochs on the same window but targeting to different
processes. This patch redesigned the RMA synchronization structure to
support per-target lock synchronization.

### CODE REVIEW

**Comment by monicasee** [comment]
test:jenkins/ch4/all

**Comment by monicasee** [comment]
test:jenkins/ch4/all

File: src/mpid/ch4/include/mpidpre.h [hunk 1]
```diff
@@ -227,45 +227,57 @@ struct MPIDI_CH4U_win_lock {
...
-    MPIDI_CH4U_win_sync_lock lock;
+} MPIDI_CH4U_win_sync_pscw_t;
+
+typedef struct MPIDI_CH4U_win_target_sync {
+    volatile int access_epoch_type;     /* NONE, LOCK. */
+    MPIDI_CH4U_win_target_sync_lock_t lock;
+} MPIDI_CH4U_win_target_sync_t;
+
+typedef struct MPIDI_CH4U_win_sync {
+    /* TODO: replace with access/exposure */
```
Comment (hajimefu) [inline src/mpid/ch4/include/mpidpre.h]: Is this TODO done?

Comment (monicasee) [inline src/mpid/ch4/include/mpidpre.h]: Oh, yes, I forgot to remove it. Thanks for the reminder.

File: src/mpid/ch4/src/ch4_impl.h [hunk 2]
```diff
@@ -551,6 +579,7 @@ do {                                                                  \
 #define MPIDI_CH4U_EPOCH_FENCE_CHECK(win, mpi_errno, stmt)              if (0) goto fn_fail;
 #define MPIDI_CH4U_EPOCH_POST_CHECK(win, mpi_errno, stmt)               if (0) goto fn_fail;
 #define MPIDI_CH4U_EPOCH_LOCK_CHECK(win, mpi_errno, stmt)               if (0) goto fn_fail;
+#define MPIDI_CH4U_EPOCH_PER_TARGET_LOCK_CHECK(win, rank, mpi_errno, stmt)  if (0) goto fn_fail;
```
Comment (hajimefu) [inline src/mpid/ch4/src/ch4_impl.h]: Definition for `MPIDI_CH4U_EPOCH_CHECK_LOCK_TYPE ` is missing.

Comment (hajimefu) [inline src/mpid/ch4/src/ch4_impl.h]: `MPIDI_CH4U_EPOCH_ORIGIN_LOCK_CHECK` is also missing.

Comment (monicasee) [inline src/mpid/ch4/src/ch4_impl.h]: These two should exist in ch4/src/ch4_impl.h.

Actually I am going to rename some of the macros as well. ORIGIN -> ACCESS, TARGET->EXP.

Comment (hajimefu) [inline src/mpid/ch4/src/ch4_impl.h]: Did you try building without `--enable-error-checking=all`? Note that you need to provide two definitions for one macro, one for `HAVE_ERROR_CHECKING` and the other for `!HAVE_ERROR_CHECKING`

Comment (hajimefu) [inline src/mpid/ch4/src/ch4_impl.h]: Also I don't know if I like "exp". "access" is 6 characters, "exposure" is 8 characters. I would propose using a full word "exposure" instead of "exp".

Comment (monicasee) [inline src/mpid/ch4/src/ch4_impl.h]: Ah, I found the missing part for !HAVE_ERROR_CHECKING. Thanks. Will add.

File: src/mpid/ch4/include/mpidpre.h [hunk 3]
```diff
@@ -227,45 +227,57 @@ struct MPIDI_CH4U_win_lock {
...
+} MPIDI_CH4U_win_lock_recvd_t;
 
-typedef struct MPIDI_CH4U_win_lock_info {
-    unsigned peer;
-    int lock_type;
-    struct MPIR_Win *win;
-    volatile unsigned done;
-} MPIDI_CH4U_win_lock_info;
+typedef struct MPIDI_CH4U_win_target_sync_lock {
+    volatile unsigned locked;   /* locked == 0 or 1 */
```
Comment (monicasee) [inline src/mpid/ch4/include/mpidpre.h]: This flag is updated by progress_engine->lock/unlock_ack callback, and waited in lock or unlock call. 

My understanding is that such update/wait can be performed by different threads, thus being dangerous with compiler optimization.

**Comment by monicasee** [comment]
test:jenkins/ch4/all

**Comment by monicasee** [comment]
test:jenkins/ch4/all

**Comment by monicasee** [comment]
test:jenkins/ch4/all

**Comment by monicasee** [comment]
test:jenkins/ch4/uti

**Comment by monicasee** [comment]
Confirmed all failing tests are unrelated.

**Comment by monicasee** [comment]
@hajimefu Can you review this patch for me ?

**Comment by hajimefu** [comment]
@monicasee sure, will do on today or tomorrow.

**Comment by hajimefu** [comment]
Reviewed through all the commits. Please take a look at each comment.

**Comment by monicasee** [comment]
@hajimefu : Thanks ! I will fix them. I have a separate comment for volatile.

File: src/mpid/ch4/include/mpidpre.h (hunk 3 above)
Comment (monicasee) [inline src/mpid/ch4/include/mpidpre.h]: Not exactly, my understanding is `volatile` is used to avoid the compiler optimization. For example, it can always read  the `locked` variable from register since no dependency is detected in the context. But the other thread who polls the progress engine can actually update it to the other value.

Comment (hajimefu) [inline src/mpid/ch4/include/mpidpre.h]: If multiple threads are sharing this `locked` variable without taking a lock (say, pthread mutex), then volatile alone isn't enough, because hardware might also optimize the instructions, so you'd need a memory barrier in such a case. If a proper lock primitive (e.g. pthread mutex) is used, then neither volatile or memory barrier is needed in your code because lock primitive will do it on behalf of you (cf. http://stackoverflow.com/a/3208140)

I think the current MPICH code falls into the latter case with a lock, at least in the global lock mode.

Comment (monicasee) [inline src/mpid/ch4/include/mpidpre.h]: This is code is actually relying on the global lock in current MPICH. That is the reason I do not use atomic operations for this flag. As long as one thread has to go into the critical section to touch the variable, I do not need additional protection or memory barrier which avoids instruction reordering. 
The only thing I am not sure is whether the compiler does optimization to keep reading `locked` from register.

**Comment by monicasee** [comment]
test:jenkins/ch4/all

File: src/mpi/coll/ired_scat.c [hunk 4]
```diff
@@ -1070,7 +1070,7 @@ int MPI_Ireduce_scatter(const void *sendbuf, void *recvbuf, const int recvcounts
                         MPI_Datatype datatype, MPI_Op op, MPI_Comm comm, MPI_Request *request)
 {
     int mpi_errno = MPI_SUCCESS;
-    int i;
+    int i ATTRIBUTE((unused));
```
Comment (monicasee) [inline src/mpi/coll/ired_scat.c]: I changed them in the latest push.

File: src/mpid/ch4/netmod/ofi/util.c [hunk 5]
```diff
@@ -340,8 +340,9 @@ static inline void MPIDI_OFI_win_unlock_done_proc(const MPIDI_OFI_win_control_t
         MPIR_Assert((int) MPIDI_CH4U_WIN(win, sync).lockall.allLocked > 0);
         MPIDI_CH4U_WIN(win, sync).lockall.allLocked -= 1;
     }
-    else
+    else {
```
Comment (monicasee) [inline src/mpid/ch4/netmod/ofi/util.c]: Yes, when disable error checking. Because MPIR_Assert(...) is just empty. So the code is like this:
if (...) {
...
} else

File: src/mpid/ch4/netmod/ofi/ofi_rma.h [hunk 6]
```diff
@@ -585,17 +585,21 @@ static inline int MPIDI_NM_mpi_put(const void *origin_addr,
     }
 
     MPIDI_CH4U_EPOCH_CHECK_SYNC(win, mpi_errno, goto fn_fail);
-    MPIDI_CH4U_EPOCH_CHECK_TARGET_SYNC(win, target_rank, mpi_errno, goto fn_fail);
     MPIDI_CH4U_EPOCH_OP_REFENCE(win);
 
+    /* Check target sync status for any target_rank except PROC_NULL. */
+    if (target_rank == MPI_PROC_NULL)
```
Comment (monicasee) [inline src/mpid/ch4/netmod/ofi/ofi_rma.h]: I changed them in the latest push.

File: src/mpid/ch4/include/mpidpre.h [hunk 7]
```diff
@@ -227,45 +227,57 @@ struct MPIDI_CH4U_win_lock {
...
+
+typedef struct MPIDI_CH4U_win_sync {
+    /* TODO: replace with access/exposure */
+    volatile int origin_epoch_type;     /* NONE, FENCE, LOCKALL, START,
+                                         * LOCK (refer to target_sync). */
+    volatile int target_epoch_type;     /* NONE, FENCE, POST. */
+
+    /* access epochs */
+    /* TODO: Can we put access epochs in union,
+     * since no concurrent epochs is allowed ? */
```
Comment (monicasee) [inline src/mpid/ch4/include/mpidpre.h]: No, I do not want to implement it in this PR. So I just note it.

File: src/mpid/ch4/include/mpidpre.h (hunk 3 above)
Comment (monicasee) [inline src/mpid/ch4/include/mpidpre.h]: I think it is safer to put `volatile ` for the variables which is only *read* within a loop of progress polling which might be fully inlined without any function calls inside, because the compiler or hardware can optimize the reading. The variables might be updated by the other thread. The function pointer triggered the updating operation can be the callback in progress engine, or potentially the event registered to network driver, thus the compiler may not be able to detect it. And the progress polling macro can be changed in future.

I removed two unneeded `volatile` (the count in `MPIDI_CH4U_win_sync_lock_t` and `MPIDI_CH4U_win_lock_recvd_t `), since they do not work as the above mode.

I also added `volatile` for the completion counters in the last commit. I have explained the reason in code comment for each `volatile`.

**Comment by monicasee** [comment]
@hajimefu : Pushed the commits again and passed Jenkins. Could you please check them ? Please see my reply for each of your previous comment. I think all comments are fixed, except the `volatile ` one. Please see my explanation in the first conversation.

File: src/mpid/ch4/include/mpidpre.h (hunk 3 above)
Comment (hajimefu) [inline src/mpid/ch4/include/mpidpre.h]: I don't get the "inlining" explanation.
1. Considering single threaded execution, compiler should be able to recognize dependencies correctly even if functions are inlined. But apparently you are talking about multi-threaded case.
2. Then in multi-threaded execution, every time a thread accesses (either read or write) the variable, it must hold a lock. Whenever read happens to the variable, the thread needs to acquire the lock beforehand. `pthread_mutex_lock` is not inlined, and even if it was inlined, it is supposed to contain both compiler/memory barrier to prevent ordering change across this lock operation. And of course once the thread takes a lock, you don't have to worry about other thread modifying the variable.

File: src/mpid/ch4/include/mpidpre.h (hunk 7 above)
Comment (hajimefu) [inline src/mpid/ch4/include/mpidpre.h]: Okay that's fine.

File: src/mpid/ch4/include/mpidpre.h (hunk 3 above)
Comment (monicasee) [inline src/mpid/ch4/include/mpidpre.h]: I think the memory barrier in pthread_mutex_lock does not avoid the compiler to optimize code by reading variable directly from register (this assembly code is already changed), it is not a memory coherence issue. 
My understanding is, the only way to avoid compiler to do such optimization is to have dependent code of that variable, or have function pointer because compiler cannot understand whether that function pointer will change the variable. However, none of these conditions is clear in the code of CH4 progress engine (e.g., `MPIDI_CH4R_PROGRESS_WHILE`). To be safe, I think it is better to have `volatile` for these variables for now.

Comment (hajimefu) [inline src/mpid/ch4/include/mpidpre.h]: Hmm, I think `pthread_mutex_lock` is supposed to work as a compiler barrier as well. It shall ["synchronize memory with respect to other threads"](http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap04.html#tag_04_12). I think this also prohibits compilers from reordering memory access across `pthread_mutex_lock`.

Comment (monicasee) [inline src/mpid/ch4/include/mpidpre.h]: This issue is not about `reordering`.  I agree that a memory barrier can avoid `reordering` optimization, and the `pthread_mutex_lock ` contains such barrier.
Here is an example of this `register caching` issue I am actually concerning.
http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0472c/CJAEGDEA.html

I also checked the assembly code of this example.
```
int cond = 1;
while(cond) {
    pthread_mutex_lock();
    pthread_mutex_unlock();
    printf("%d", cond);
}
```
The compiler (just tried GCC) does not do the `register caching` or `replacing with constant` for `cond`, is not because of the barrier thing, it is actually because it cannot understand the behavior of external functions such as `pthread_mutex_lock`. I tried to use just `extern void dummy();` to replace these functions, it does the same thing -- the `cond` is reloaded from memory for `printf` line.

Although the current progress engine code coincidently calls external functions, thus should not fall into the issue I am concerning. I am still standing on the side of using `volatile` in the RMA code, just to be safe in case one changes the progress engine code  in future.

Comment (hajimefu) [inline src/mpid/ch4/include/mpidpre.h]: I would think the order of lock and unlock in your example should be the other way around, but anyway.

Same discussion for reordering applies to the register caching as well. Since `pthread_mutex_lock/unlock` is supposed to synchronize memory with respect to other threads, compiler cannot cache values (that may be shared with other threads) in the register if these calls appear.

>The compiler (just tried GCC) does not do the register caching or replacing with constant for cond, is not because of the barrier thing, it is actually because it cannot understand the behavior of external functions such as pthread_mutex_lock.

This is also an important point. Beside the specification of `pthread_mutex_lock` above, lock primitives are most likely implemented as external library functions, so compilers cannot cache values anyway.

To summarize, I strongly believe we don't need `volatile` as long as pthread mutex is used.

Comment (monicasee) [inline src/mpid/ch4/include/mpidpre.h]: I agree your point that as long as the `pthread` lock functions are used, the compiler will not optimize the code by register caching it, thus we do not need `volatile` per variable. 

However, if other lock mechanism is used in future, such as Halim's `izem`, which can be fully inlined with zero function call, then the above assumption cannot work. There is actually a plan to add the configure option to allow user to specify the thread CS implementation. 

All I am saying is, the lock theory itself does not guarantee no register caching, it is implementation dependent.  `volatile` is the only precise way to avoid  register caching, which solves the specific problem of the RMA variables.

**Comment by monicasee** [comment]
Manually fixed a rebase conflict.

**Comment by monicasee** [comment]
test:jenkins/ch4/all

**Comment by monicasee** [comment]
test:jenkins/ch4/ofi
